core/modules/node/node.module

```diff
@@ -3056,6 +3057,47 @@ function _node_query_node_access_alter($query, $type) {
 
 }
 
+/**
+ * Exclude nodes from the search query where the node has a hidden path
+ * and the current user doesn't have permission to bypass the restriction.
+ */
+function _node_search_query_alter(QueryAlterableInterface $query) {
+  global $user;
+
+  if ($user->uid == 1) {
+    return;
+  }
+
+  if (in_array('view hidden paths', user_role_permissions($user->roles))) {
```

Comment:
Curious why not use `user_access()` here.

Reply:
Good point. Not sure why I complicated it. 🙂

Reply:
Fixed @klonos.